File: NBE_5_7_2_02TEAM/src/main/java/io/twogether/nbe_5_7_2_02team/member/util/mapper/MemberMapper.java
package io.twogether.nbe_5_7_2_02team.member.util.mapper;

import io.twogether.nbe_5_7_2_02team.member.domain.Member;
import io.twogether.nbe_5_7_2_02team.member.dto.response.MemberCreateResponse;
import io.twogether.nbe_5_7_2_02team.member.dto.response.MyPageResponse;
import io.twogether.nbe_5_7_2_02team.member.dto.response.MyPageResponse.PostSummary;
import io.twogether.nbe_5_7_2_02team.post.domain.Post;

import java.util.List;

public class MemberMapper {

    public MemberMapper() {}

    public static MemberCreateResponse toMemberCreateResponse(Member member) {
        return MemberCreateResponse.builder()
                .id(member.getId())
                .email(member.getEmail())
                .name(member.getName())
                .profileImage(member.getProfileImage())
                .job(member.getJob())
                .course(member.getCourse())
                .githubId(member.getGithubId())
                .build();
    }

    public static MyPageResponse toMyPageResponse(
            Member member,
            List<Post> posts,
            Long followerCount,
            Long followingCount,
            boolean following,
            boolean owner) {

        List<PostSummary> postSummaries =
                posts.stream()
                        .map(
                                post ->
                                        MyPageResponse.PostSummary.builder()
                                                .postId(post.getId())
                                                .title(post.getTitle())
                                                .build())
                        .toList();

        return MyPageResponse.builder()
                .id(member.getId())
                .email(member.getEmail())
                .name(member.getName())
                .job(member.getJob())
                .course(member.getCourse())
                .profileImage(member.getProfileImage())
                .posts(postSummaries)
                .followerCount(followerCount)
                .followingCount(followingCount)
                .following(following)
                .owner(owner)
                .build();
    }
}

File: NBE_5_7_2_02TEAM/src/main/java/io/twogether/nbe_5_7_2_02team/member/util/mapper/MemberMapper.kt
package io.twogether.nbe_5_7_2_02team.member.util.mapper

import io.twogether.nbe_5_7_2_02team.member.domain.Member
import io.twogether.nbe_5_7_2_02team.member.dto.response.MemberCreateResponse
import io.twogether.nbe_5_7_2_02team.member.dto.response.MemberUpdateResponse
import io.twogether.nbe_5_7_2_02team.member.dto.response.MyPageResponse
import io.twogether.nbe_5_7_2_02team.member.dto.response.MyPageResponse.PostSummary
import io.twogether.nbe_5_7_2_02team.post.domain.Post

fun Member.toMemberCreateResponse(): MemberCreateResponse =
    MemberCreateResponse(
        id = this.id!!,
        email = this.email,
        name = this.name,
        profileImage = this.profileImage,
        job = this.job,
        course = this.course,
        githubId = this.githubId,
    )

fun Member.toMemberUpdateResponse(
    followerCount: Long,
    followingCount: Long,
): MemberUpdateResponse =
    MemberUpdateResponse(
        id = this.id!!,
        name = this.name,
        profileImage = this.profileImage,
        followerCount = followerCount,
        followingCount = followingCount,
        following = false,
        owner = true,
    )

fun Member.toMyPageResponse(
    posts: List<Post>,
    followerCount: Long,
    followingCount: Long,
    following: Boolean,
    owner: Boolean,
): MyPageResponse =
    MyPageResponse(
        id = this.id!!,
        email = this.email,
        name = this.name,
        job = this.job,
        course = this.course,
        profileImage = this.profileImage,
        posts = posts.map { it.toPostSummary() },
        followerCount = followerCount,
        followingCount = followingCount,
        following = following,
        owner = owner,
    )

fun Post.toPostSummary(): PostSummary =
    PostSummary(
        postId = this.id!!,
        title = this.title,
    )